common/src/main/java/org/apache/atlas/service/FeatureFlagStore.java

```diff
@@ -2,128 +2,379 @@
+        
+        // Cap the maximum delay to prevent extremely long waits (e.g., 30 seconds max)
+        long maxDelayMs = 30000L;
+        return Math.min(backoffDelay, maxDelayMs);
+    }
+
+    public static boolean isTagV2Enabled() {
+        return !evaluate(FeatureFlag.ENABLE_JANUS_OPTIMISATION.getKey(), "false"); // Default value is false, if the flag is present or has any other value it's treated as enabled
     }
 
     public static boolean evaluate(String key, String expectedValue) {
-        boolean ret = false;
-        try{
-            if (redisService == null) {
-                return false;
+        return StringUtils.equals(getFlag(key), expectedValue);
+    }
+
+    public static String getFlag(String key){
+        try {
+            if (!isValidFlag(key)) {
+                LOG.warn("Invalid feature flag requested: '{}'. Only predefined flags are allowed", key);
+                return null;
+            }
+
+            FeatureFlagStore instance = getInstance();
+            if (instance == null) {
+                LOG.warn("FeatureFlagStore not initialized, cannot get flag: {}", key);
+                return getDefaultValue(key);
             }
-            if (StringUtils.isEmpty(key) || StringUtils.isEmpty(expectedValue))
-                return ret;
-            String value = redisService.getValue(addFeatureFlagNamespace(key));
-            ret = StringUtils.equals(value, expectedValue);
+
+            return instance.getFlagInternal(key);
         } catch (Exception e) {
-            LOG.error("Error evaluating feature flag '{}' with expected value '{}'", key, expectedValue, e);
-            return ret;
+            LOG.error("Error getting feature flag '{}'", key, e);
+            return getDefaultValue(key); // Always return something
         }
-        return ret;
     }
 
-    public static String getFlag(String key) {
-        String ret = "";
-        try{
-            if (redisService == null) {
-                return "";
-            }
-            if (StringUtils.isEmpty(key))
-            {
-                return ret;
+    private static String getDefaultValue(String key) {
+        FeatureFlag flag = FeatureFlag.fromKey(key);
+        return flag != null ? String.valueOf(flag.getDefaultValue()) : "false";
+    }
+
+    private static boolean isValidFlag(String key) {
+        return FeatureFlag.isValidFlag(key);
+    }
+
+    String getFlagInternal(String key) {
+        if (!initialized) {
+            LOG.warn("FeatureFlagStore not fully initialized yet, attempting to get flag: {}", key);
+            throw new IllegalStateException("FeatureFlagStore not initialized");
+        }
+
+        if (redisService == null) {
+            LOG.error("RedisService is null - this should never happen after proper initialization");
+            throw new IllegalStateException("RedisService is not available");
+        }
+
+        if (StringUtils.isEmpty(key)) {
+            return "";
```

Comment:
Let us throw exception here as well, I do not see any reason why we should silenly ignore empty key evaluations

Reply:
Okay